blog/content/post/2022-04-07-fobnail_attestation_phase.md

```diff
@@ -0,0 +1,93 @@
+---
+title: Fobnail Token - platform attestation
+abstract: 'The Fobnail Token is an open-source hardware USB device that helps to
+           determine the integrity of the system. The purpose of this blog post
+           is to present the development progress of this project. This phase
+           was focused on attestation.'
+cover: /covers/usb_token.png
+author: krystian.hebel
+layout: post
+published: true
+date: 2022-04-07
+archives: "2022"
+
+tags:
+  - fobnail
+  - firmware
+  - tpm
+  - attestation
+categories:
+  - Firmware
+  - Security
+
+---
+
+# About the Fobnail Token project
+
+The Fobnail Token is a project that aims to provide a reference architecture for
+building offline integrity measurement servers on the USB device and clients
+running in Dynamically Launched Measured Environments (DLME). It allows the
+Fobnail owner to verify the trustworthiness of the running system before
+performing any sensitive operation. This project was founded by [NlNet
+Foundation](https://nlnet.nl/). More information about the project can be found
+in the [Fobnail documentation](https://fobnail.3mdeb.com/). Also, make sure to
+read other posts related to this project by visiting
+[fobnail](https://www.blog.3mdeb.com/tags/fobnail/) tag.
+
+# Attestation
+
+## Architecture
+
+## Implementation
+
+# Demo
+
+[![asciicast](https://asciinema.org/a/OJ1YWyKhexSztmbfNVS79eGbo.svg)](https://asciinema.org/a/OJ1YWyKhexSztmbfNVS79eGbo?speed=1)
+
+# Running Fobnail on real hardware
+
+During early development, we used nRF52 as a device for running Fobnail
+firmware. However, due to problems with USB, we started running Fobnail as a
+Linux application during the previous phase. The time has come to fix this.
+
+## Fixing USB
+
+Many times when Fobnail was plugged into USB, it didn't work properly. This
+problem has been described
+[here](https://github.com/fobnail/usbd-ethernet/issues/2). We have searched thru
+issues and PRs of the libraries we use and updated them to their latest
+versions, but it didn't help, so we started looking for the issue in our own
+code.
+
+The direct cause of USB failure was a too big delay between USB interrupts, up
+to 85 ms, which occurred right after USB initialization. At first, we tried
```

Comment:
@arturkow2000 how you managed to realize that? Please tell the story.

Reply:
Interrupt delays were actually one of the first things I checked. It was much harder to find out why it happens because I wrongly assumed that critical sections were the cause.

Reply:
@arturkow2000 this blog post should tell a story, please rephrase sentence to explain why you checked interrupts as first and what you found.